File: vs_xam/baby-flashcards/baby-flashcards/ViewLogic/JumbleFlowLayout.cs
using System;
using UIKit;
using CoreGraphics;
using Foundation;
using System.Collections.Generic;

namespace babyflashcards
{
	public class JumbleFlowLayout : UICollectionViewFlowLayout
	{
		private Dictionary<int, PosStore> _lastPositions;

		public JumbleFlowLayout ()
		{
			_lastPositions = new Dictionary<int, PosStore> ();
		}

		public override void PrepareLayout ()
		{
			base.PrepareLayout ();
			var cellCount = (int)CollectionView.NumberOfItemsInSection (0);

			for (int i = 0; i < cellCount; i++) {
				var itemIndexPath = NSIndexPath.FromItemSection (i, 0);
				var item = (CollectionView.DataSource as DeckCollectionViewController).Tiles [itemIndexPath.Row];
				var attributes = LayoutAttributesForItem (itemIndexPath);
				if (!_lastPositions.ContainsKey (item.ThumbId))
					_lastPositions.Add (item.ThumbId, new PosStore ());
				_lastPositions [item.ThumbId].StorePos (attributes.Center);

			}
		}

		public override void PrepareForAnimatedBoundsChange (CGRect oldBounds)
		{
			var cellCount = (int)CollectionView.NumberOfItemsInSection (0);

			for (int i = 0; i < cellCount; i++) {
				var itemIndexPath = NSIndexPath.FromItemSection (i, 0);
				var item = (CollectionView.DataSource as DeckCollectionViewController).Tiles [itemIndexPath.Row];
				var attributes = LayoutAttributesForItem (itemIndexPath);
				if (!_lastPositions.ContainsKey (item.ThumbId))
					_lastPositions.Add (item.ThumbId, new PosStore ());
				_lastPositions [item.ThumbId].StorePos (attributes.Center);
			}
			base.PrepareForAnimatedBoundsChange (oldBounds);

		}

		public override UICollectionViewLayoutAttributes InitialLayoutAttributesForAppearingItem (NSIndexPath itemIndexPath)
		{
			var attributes = base.InitialLayoutAttributesForAppearingItem (itemIndexPath);

			var item = (CollectionView.DataSource as DeckCollectionViewController).Tiles [itemIndexPath.Row];

			attributes.Alpha = 1;
			attributes.Center = _lastPositions [item.ThumbId].LastPos;

			return attributes;
		}

		public override UICollectionViewLayoutAttributes FinalLayoutAttributesForDisappearingItem (NSIndexPath itemIndexPath)
		{
			var attributes = base.FinalLayoutAttributesForDisappearingItem (itemIndexPath);
			var cell = CollectionView.CellForItem (itemIndexPath) as DeckViewCell;
			cell.Layer.Hidden = true;
			cell.Layer.RemoveAllAnimations ();

			return attributes;
		}

		private class PosStore
		{
			private CGPoint _lastPos = CGPoint.Empty;
			private CGPoint _nextPos = new CGPoint (0, 0);

			public void StorePos (CGPoint pos)
			{
				_lastPos = _nextPos;
				_nextPos = pos;
			}

			public CGPoint LastPos {
				get {
					if (_lastPos == CGPoint.Empty)
						return _nextPos;
					return _lastPos;
				}
			}

			public CGPoint NextPos {
				get {
					return _nextPos;
				}
			}

		}

	}


}

File: xcode/Scratch1/Scratch1/JumbleFlowLayout.swift
//
//  JumbleFlowLayout.swift
//  baby-flashcards
//
//  Created by Michael Ishmael on 13/05/2016.
//  Copyright © 2016 66Bytes. All rights reserved.
//

import Foundation
import UIKit

class JumbleFlowLayout : UICollectionViewFlowLayout {
    
    var _lastPositions = [Int: PosStore]()
    var _firstRun = false;
    var _hold = false
    
    override func prepareLayout() {
        super.prepareLayout()
        if(!_firstRun){
            calculatePositions()
            _firstRun = true;
        }
        
    }
    
    private func calculatePositions(){
        if let cellCount =  collectionView?.numberOfItemsInSection(0) {
            for i in 0...(cellCount-1) {
                let itemIndexPath = NSIndexPath(forItem: i, inSection: 0)
                let item = (collectionView?.dataSource as! DeckCollectionViewController!).tiles[itemIndexPath.row]
                let attributes = self.layoutAttributesForItemAtIndexPath(itemIndexPath)
                
                if _lastPositions[item.thumbId] == nil {
                    _lastPositions[item.thumbId] = PosStore()
                }
                if let centerPos = attributes?.center {
                    _lastPositions [item.thumbId]?.storePos (centerPos);
                }
            }
        }
    }
    
    internal func doneRotation(){
        _lastPositions = [Int: PosStore]()
        calculatePositions()
        _hold = false;
    }
    
    override func prepareForCollectionViewUpdates(updateItems: [UICollectionViewUpdateItem]) {
        calculatePositions()
        super.prepareForCollectionViewUpdates(updateItems)
    }
    
    override func invalidateLayout() {
        super.invalidateLayout()
        //calculatePositions()
    }
    
    override func prepareForAnimatedBoundsChange(oldBounds: CGRect) {
        //calculatePositions()
        super.prepareForAnimatedBoundsChange(oldBounds)
        _hold = true;
    }
    
    override func initialLayoutAttributesForAppearingItemAtIndexPath(itemIndexPath: NSIndexPath) -> UICollectionViewLayoutAttributes? {
        
        let attributes = super.initialLayoutAttributesForAppearingItemAtIndexPath(itemIndexPath)
        
        if(_hold) { return attributes};
        
        let item = (self.collectionView?.dataSource as! DeckCollectionViewController).tiles[itemIndexPath.row]
        attributes?.alpha = 1
        attributes?.center = (_lastPositions[item.thumbId]?.lastPos)!
        
        return attributes
        
    }
    
    override func finalLayoutAttributesForDisappearingItemAtIndexPath(itemIndexPath: NSIndexPath) -> UICollectionViewLayoutAttributes? {
        
        let attributes = super.initialLayoutAttributesForAppearingItemAtIndexPath(itemIndexPath)
        
        if(_hold) { return attributes};
        
        let cell = self.collectionView?.cellForItemAtIndexPath(itemIndexPath) as! DeckViewCell
        cell.layer.hidden = true;
        cell.layer.removeAllAnimations()
        
        return attributes
    }
    
    
    class PosStore
    {
        var _lastPos:CGPoint? = nil
        var nextPos:CGPoint = CGPoint (x: 0, y: 0);
    
        func storePos(pos:CGPoint) {
            _lastPos = nextPos;
            nextPos = pos;
        }
        
        var lastPos: CGPoint {
            get{
                if(_lastPos == nil){
                    return nextPos;
                }
                return _lastPos!;
            }
        }
        
    }
    
}
